sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -111,13 +111,22 @@ case class AdaptiveSparkPlanExec(
     DisableUnnecessaryBucketedScan
   ) ++ context.session.sessionState.queryStagePrepRules
 
+  // This list rules are applied between queryStagePreparationRules and estimate physical plan cost
+  // so that we can support introduce extra shuffle
```

Comment:
I'm thinking about a more efficient way that runs `EnsureRequirements` only once
```
def queryStagePreparationRules(optimizeSkewJoin: Boolean) = {
  val rules = Seq(
    RemoveRedundantProjects,
    OptimizeSkewedJoin,
    ...
  )
  if (optimizeSkewJoin) rules else rules.filterNot(_ != OptimizeSkewedJoin)
}
```